src/channel.c

```diff
@@ -1894,19 +1894,10 @@ libssh2_channel_read_ex(LIBSSH2_CHANNEL *channel, int stream_id, char *buf,
                         size_t buflen)
 {
     int rc;
-    unsigned long recv_window;
 
     if(!channel)
         return LIBSSH2_ERROR_BAD_USE;
 
-    recv_window = libssh2_channel_window_read_ex(channel, NULL, NULL);
```

Comment:
When 03ca902 was added, the channel window size was not precisely tracked, and it drifted resulting sometimes in deadlocks. As a workaround, reading methods were changed to reserve enough space in the windows for its operation and that fixed the deadlock issue, but broke the SSH flow control.

I added precise tracking of the window size in cdeef54967ed5b7d5bd8fa6da5851aa3d173faa0, making all of those workarounds useless.

Now a simple algorithm that sends window adjustment packets when the window size goes below a threshold can be used safely.